File: src/main/java/destiny/utils/LocaleComparator.java
package destiny.utils;



import java.io.Serializable;
import java.util.Comparator;
import java.util.Locale;
import java.util.Objects;

/**
 * Created by smallufo on 2014-03-05.
 */
public class LocaleComparator implements Comparator<Locale>, Serializable {
  private final Locale locale;

  public LocaleComparator(Locale locale) {
    this.locale = locale;
  }

  @Override
  public int compare(Locale locale1, Locale locale2) {
    int result;
    if (Objects.equals(locale1, locale2))
      result = 0;
    else if (Objects.equals(locale1, locale))
      result = -1;
    else if (Objects.equals(locale2, locale))
      result = 1;
    else if (Objects.equals(locale1.getLanguage(), locale.getLanguage()))
      result = -1;
    else if (Objects.equals(locale2.getLanguage(), locale.getLanguage()))
      result = 1;
    else if (Objects.equals(Locale.ENGLISH.getLanguage(), locale1.getLanguage()))
      result = -1;
    else if (Objects.equals(Locale.ENGLISH.getLanguage(), locale2.getLanguage()))
      result = 1;
    else
      result = 1;
    return result;
  }
}

File: src/main/java/destiny/tools/LocaleComparator.kt
/**
 * Created by smallufo on 2016-02-22.
 */
package destiny.tools

import java.io.Serializable
import java.util.*
import kotlin.Comparator

class LocaleComparator(private val locale: Locale) : Comparator<Locale>, Serializable {

  override fun compare(locale1: Locale, locale2: Locale): Int {
    return when {
      locale1 == locale2 -> 0
      locale1 == locale -> -1
      locale2 == locale -> 1
      locale1.language == locale.language -> -1
      locale2.language == locale.language -> 1
      Locale.ENGLISH.language == locale1.language -> -1
      Locale.ENGLISH.language == locale2.language -> 1
      else -> 1
    }
  }
}
